engine/access/rpc/backend/backend_test.go

```diff
@@ -393,6 +395,143 @@ func (suite *Suite) TestGetLatestProtocolStateSnapshot_HistoryLimit() {
+		epochBuilder := unittest.NewEpochBuilder(suite.T(), state)
+		// build epoch 1
+		// Blocks in current State
+		// P <- A(S_P-1) <- B(S_P) <- C(S_A) <- D(S_B) |setup| <- E(S_C) <- F(S_D) |commit|
+		epochBuilder.
+			BuildEpoch().
+			CompleteEpoch()
+
+		// get heights of each phase in built epochs
+		epoch1, ok := epochBuilder.EpochHeights(1)
+		require.True(suite.T(), ok)
+
+		// setup AtBlockID and AtHeight mock returns for State
+		for _, height := range epoch1.Range() {
+			snap := state.AtHeight(height)
+			blockHead, err := snap.Head()
+			suite.Require().NoError(err)
+
+			suite.state.On("AtHeight", height).Return(snap)
+			suite.state.On("AtBlockID", blockHead.ID()).Return(snap)
+		}
+
+		// Take snapshot at height of block D (epoch1.heights[2]) for valid segment and valid snapshot
+		snap := state.AtHeight(epoch1.Range()[2])
+		blockHead, err := snap.Head()
+		suite.Require().NoError(err)
+		suite.state.On("Final").Return(snap).Once()
+
+		params := suite.defaultBackendParams()
+		params.MaxHeightRange = TEST_MAX_HEIGHT
+
+		backend, err := New(params)
+		suite.Require().NoError(err)
+
+		// query the handler for the latest finalized snapshot
+		bytes, err := backend.GetProtocolStateSnapshotByBlockID(context.Background(), blockHead.ID())
+		suite.Require().NoError(err)
+
+		// we expect the endpoint to return the snapshot at the same height we requested
+		expectedSnapshotBytes, err := convert.SnapshotToBytes(snap)
+		suite.Require().NoError(err)
+		suite.Require().Equal(expectedSnapshotBytes, bytes)
+	})
+}
+
+// TestGetProtocolStateSnapshotByBlockID_Non_Finalized_Blocks tests our GetProtocolStateSnapshotByBlockID RPC endpoint
+// where non finalized block is added to state
+func (suite *Suite) TestGetProtocolStateSnapshotByBlockID_NonFinalizedBlocks() {
+	identities := unittest.CompleteIdentitySet()
+	rootSnapshot := unittest.RootSnapshotFixture(identities)
+	util.RunWithFullProtocolState(suite.T(), rootSnapshot, func(db *badger.DB, state *bprotocol.ParticipantState) {
+		epochBuilder := unittest.NewEpochBuilder(suite.T(), state)
+		// build epoch 1
+		// Blocks in current State
+		// P <- A(S_P-1) <- B(S_P) <- C(S_A) <- D(S_B) |setup| <- E(S_C) <- F(S_D) |commit|
+		epochBuilder.BuildEpoch()
+
+		// get heights of each phase in built epochs
+		epoch1, ok := epochBuilder.EpochHeights(1)
+		require.True(suite.T(), ok)
+
+		// Take snapshot at height of last block of epoch
+		snap := state.AtHeight(epoch1.CommittedFinal)
+		snapHead, err := snap.Head()
+		suite.Require().NoError(err)
+		newBlock := unittest.BlockWithParentFixture(snapHead)
+		ctx := context.Background()
+		// Adding new non finalized block to state
+		err = state.Extend(ctx, newBlock)
+		suite.Require().NoError(err)
+
+		// setup AtHeight and AtBlockID mock returns for State
+		suite.state.On("AtHeight", newBlock.Header.Height).Return(state.AtHeight(newBlock.Header.Height))
+		suite.state.On("AtBlockID", newBlock.ID()).Return(state.AtBlockID(newBlock.ID()))
```

Comment:
wouldn't this result in it appearing as finalized? if this is correct, can you expand on the comment explaining what's happening and the expected behavior?